### PR TITLE
refactor(stages): Adding StageConfigWrapper for cleaner StageConfigs

### PR DESCRIPTION
**Pull request by alanmquach**
When stage fields are updated, instead of mutating props directly and force updating, `StageConfig` components should actually invoke a callback with the changes.

Before:
```
this.props.stage.variables = variables;
this.props.stageFieldUpdated();
this.forceUpdate();
```

After
```
this.props.updateStageField({variables});
```

`StageConfigWrapper` exists purely to perform the `this.forceUpdate()` because we wouldn't want all the `StageConfig` components to have to make that call (or even be aware of it). The reason the force update is needed is because `$scope.stage` is updated above the render root. Once that can be encapsulated in the state of some parent component (when we migrate it from angular to react), the force update and the wrapper component should no longer be needed.

### CODE REVIEW

File: app/scripts/modules/core/src/pipeline/config/stages/StageConfigWrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import * as React from 'react';
+
+import { IStageConfigProps } from 'core/pipeline';
+
+export interface IStageConfigWrapperProps extends IStageConfigProps {
+  component: typeof React.Component;
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/pipeline/config/stages/StageConfigWrapper.tsx]:
```suggestion
  component: React.ComponentType<any>;
```

Comment (christopherthielen) [reply app/scripts/modules/core/src/pipeline/config/stages/StageConfigWrapper.tsx]: Prefer `ComponentType`. i think this syntax is correct, but who knows

**Comment by christopherthielen** [comment]
Very like.
Does it work?
Are there other stages that need to be updated in the same way as evaluate variables?

**Comment by erikmunson** [comment]
❤️ ❤️ ❤️

**Comment by jrsquared** [comment]
I love this. Great idea/pattern for us to follow.

**Comment by alanmquach** [comment]
It works indeedy. Yes, I think all the react stages do this, but there are only a handful of them.

I plan to update them next. Wanted to get some feedback on this before doing the bulk change.

File: app/scripts/modules/core/src/pipeline/config/stages/StageConfigWrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/* This wrapper component exists so that StageConfig components don't have to manually
+ * call this.forceUpdate() after updating a stage field. The only reason it is currently
+ * needed is because the stage config object is managed in angular, above the render root.
+ * When it is eventually migrated to react, it will simply be part of the state of some
+ * parent component and this wrapper can be removed.
+ */
+
+export class StageConfigWrapper extends React.Component<IStageConfigWrapperProps> {
+  public render() {
+    const { component: StageConfig, updateStageField, ...otherProps } = this.props;
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/pipeline/config/stages/StageConfigWrapper.tsx]: I don't think `component` is used?

Comment (alanmquach) [reply app/scripts/modules/core/src/pipeline/config/stages/StageConfigWrapper.tsx]: It destructures `this.props.component` into `StageConfig`, which is used.

Comment (anotherchrisberry) [reply app/scripts/modules/core/src/pipeline/config/stages/StageConfigWrapper.tsx]: ahhh, I didn't know you could rename variables in a destructuring call! Thanks!

**Pull request merged**